### DISCUSSION TITLE
Keep older versions of extensions

### DISCUSSION THREAD

**Discussion by arthuro555** [4ian/GDevelop#2809] [Enhancement ideas (and their technical discussions)]
There is no discussions for GDevelopApp/GDevelop-extensions so I am opening one here.

I think the extensions repository should support having multiple versions of extensions, or at least all major versions to be able to rollback from a major release with breaking changes that are too hard to deal with at the moment.

**Comment by Bouh** [comment]
Isn't possible to fetch the git history of this repo to get a specific version of an extension from the app with a dropdown list?

**Reply by arthuro555** [reply]
It is technically possible I guess but I'm pretty sure the autobuild is hosting everything on a separate server, not on the github repository, making the built extension repository not linked to the github repository. It would also add a lot of complexity and dependencies to the IDE, where it is tbh not that hard to just make a copy of the extension on major releases of it.

**Comment by 4ian** [comment]
Do we have an example of use cases where this is a problem? For now I only see the problem being "I wanted to update the extension, turns out it's breaking my game".
Which is a valid problem, but maybe we're thinking about the wrong solution? 
As a GDevelop user, I don't care about your "versioning", I just want something that works. Maybe I just want an "undo" button to unbrick my game. Maybe we should just display a warning "This extension has changed, you have to be prepared to adapt your game" so that I know I have some homework to do (and ideally there is a list of what changed).
Maybe we should suggest to update to a lower version as you suggest (but do we really want to encourage that? We're not npm, we're just a library of extensions and we want to encourage people using the latest version. Do we really want to maintain 2, 3 or 4 major versions for each extension? Can't we just keep backward compatibility for as long as possible?).

I would strongly push against keeping older versions for now, and find better solutions for compatibility and think more deeply about what is the real problem :)

**Reply by arthuro555** [reply]
Ok, that's fine with me. That means we have to be extra careful designing the API though as we don't have as much room to "experiment" 😬

**Reply by 4ian** [reply]
As I don't want to limit you too much, I think you can still experiment, but we'll clearly mark a big " (experimental)" to the extension if we add it to the library (with a version number starting with 0.x.y).

In the future, we can open up the extensions library much more and make more categories of extensions: 
- recommended (battle proof and easy to use extension)
- advanced (needs knowledge in something particular, but still fairly tested)
- experimental (you might need to entirely remove the extension from your game as it's not officially supported if something goes wrong).

and we could make sure that anything in the experimental category has a checkbox "I understand this is a preview and might not turn into a supported extension or could be removed from the library in the future" :)
